hercules-ci-api/src/Hercules/Formats/CacheKeys/Keys.hs

```diff
@@ -0,0 +1,33 @@
+module Hercules.Formats.CacheKeys.Keys where
+
+import           Prelude
+
+import           Data.Aeson
+import           Data.Foldable
+import           Data.Text                      ( Text )
+
+-- | Credentials and public keys for a cache.
+data Keys = Keys
+ { signingKeys :: [Text]
+ , pullToken :: Maybe Text
+ , publicKeys :: [Text]
+ }
```

Comment:
This way, the whole `pullToken` field is optional, instead of putting a nullable value in a required field.

Reply:
```haskell
data Foo = Foo { a :: Maybe String, b :: String } 
  deriving (Generic, Show)
instance FromJSON Foo where 
  parseJSON = genericParseJSON (defaultOptions { omitNothingFields = True })
instance ToJSON Foo where 
  toJSON = genericToJSON (defaultOptions { omitNothingFields = True })


λ decode "{\"b\":\"c\"}" :: Maybe Foo
Just (Foo {a = Nothing, b = "c"})

λ encode (Foo Nothing "c")
"{\"b\":\"c\"}"
```

there's probably a way with deriving via but didn't bother.

Reply:
Ah, yes, for Maybe it can be done, but for `[]` it can't. So yeah that would work. I decided against omitting `signingKeys` and `publicKeys` because they are too important. A public key is basically required in any case and when omitting the signing key, it's good to be explicit about it. `pullToken` only applies to private caches, so I made that field entirely optional for conciseness.
What a coincidence that it works out this way.
Not sure if `genericToJSON` is an improvement, now that we have the expanded version, because it obscures the path to optional list attributes.

Reply:
> Not sure if genericToJSON is an improvement, now that we have the expanded version, 

It's way less code to read and thus change.

> obscures the path to optional list attributes.

Not sure I understand, lists are not optional. `omitNothingFields` hides only `null` in both directions.

Reply:
Considering that the implementations are equivalent, I was comparing maintainability. Current code seems easier to modify.

> It's way less code to read and thus change

Aeson's generic functions don't help with reasoning about compatibility nor with providing the logic that is required for compatibility. I've previously had to expand the generic bits in order to add in compatibility logic. With the current code that work has already been done, so it will be less effort to modify this type in a compatible way.

Reply:
> Aeson's generic functions don't help with reasoning about compatibility nor with providing the logic that is required for compatibility

That's the role of tests, the implementation can still be simple. I'd rather use generics and write tests that assert JSON is correct on CI than rely on reading long implementations of handcrafted JSON with combinators and reason through each time.

> I've previously had to expand the generic bits in order to add in compatibility logic. 

I do remember that, but I can't remember what kind it was, but you can easiy add new fields that are optional with `Maybe` and you can rename them with `constructorTagModifier`.

My main worry here is that this reasoning is dangerous, you end up writing all JSON parsers by hand then as you might want to change them later in some weird way. 

If you really have no choice but to not use generics that's fine, but "one day we might not" is a slippery slope.

Reply:
I agree that you don't want to start out like this, but going back to generic doesn't seem effective in this case.

Reply:
I argued that once you do have to skip generics you obviously would, but we don't have to?